### ISSUE TITLE
Only support Node v4.0.0 and up

### ISSUE THREAD

**Issue by mickhansen** [sequelize/sequelize#5205]
To lower CI times for increased productivity and sense of well being.

**Comment by alitaheri** [comment]
I agree with this :+1:

**Comment by Verdier** [comment]
Agree :+1:  and project could use ES6 features such as `class` or *even* `co`...?! :battery:

**Comment by mickhansen** [comment]
@Verdier We're likely not moving to generators internally, since ideally we'd still have unofficial support for older v8 versions with less features.

We plan to add class support yes, however we still need to support the old sequelize.define :)

**Comment by felixfbecker** [comment]
And dont forget how much the codebase would benefit from arrow functions (also the docs).
Maybe this would also be the time to refactor the code so that "instances" are actually instances of models? E.g. `Sequelize.prototype.define` would return an anonymous class that extends the Model class, and "instances" will actually be instances of that class.

**Comment by felixfbecker** [comment]
And optionally, you can pass `define` you own class as the first argument as long as it extends `Model`, so you can actually define getters/setters as real getters/setters and customize your model with everything, even `toJson()`.

**Comment by mickhansen** [comment]
@felixfbecker yeah the plan is to change it so instances are instanceas of models, we need to change a lot of model checks though.

Plan is to support both ES6 classes and define.
As for arrow functions, definitely :) Away with var self = this

**Comment by felixfbecker** [comment]
Then in the very distant ES7 future instead of `define` you could use a decorator on your model class
```js
@attributes({
  ...
})
@tableName('...')
class Test extends Sequelize.Model { ... }
```

**Comment by mickhansen** [comment]
@felixfbecker yeah, there's another issue for ES6 class support, you can add your help and opinons here.
For ES6 we're probably looking at something like:

```js
class Test extends Sequelize.Model {
  static tableName () => 'test'
  static attributes () => ({
    firstName: Sequelize.STRING
  })
}
```

**Comment by felixfbecker** [comment]
@mickhansen No, please please please. Don't do this. Thats what I hate the most about PHP ORMs, that just because the language is so inflexible you always have to implement these methods that do nothing but return the configuration. But Javascript _is_ flexible. You can assign stuff on classes. You can build classes dynamically. Think about it, currently `define` is basically a class factory. You pass it arguments _beforehand_ that define how this new model will map to the database.

With a class based approach, this would map perfectly to decorators - which are nothing but a function that take a class (or anything) and _decorate_ it with additional info. Until decorators are here, this would mean that you simply pass define the class as a first argument, and the additional info as second argument.

I think static methods should have nothing to do with that additional info. You could add your own static methods like `findByMyCustomAttribute()`, just like you could add your own instance methods and getters and setters. It also doesnt make sense - the table name is never dynamic. It should be defined once and not returned by a method. Btw, your example doesnt work in ES6, as there are no arrow methods on classes. You would have to write a full method with a return statement.

Just my 0.02€

**Comment by mickhansen** [comment]
@felixfbecker the ES6 class system does now allow static properties defined in the class method.
Decorators are not here for everyone yet, and we'd love to support ES6 classes fully still.

What would you suggest, with a code example, instead of going full ES6 classes?
I suggested this syntax because that's how React does it and i find that to be okay, but i certainly welcome any suggestions :)

The implementations however needs to be mappable from sequelize.define()

**Comment by Verdier** [comment]
@mickhansen @laggingreflex 

Just to let you know, here is what we use today: (ES7 + Decorators + Babel)
```js
@model()
class MyModel {
 @attribute(Sequelize.INTEGER)
 @validate({ min: 0 })
 hours;
 
 @hook('afterValidate')
 async validate() {
 }

 ...
}
```

But that's not a solution for ES6 without transpiler... For ES6 we could have something like that:

```js
class MyModel extends Sequelize.ModelFactory({
 hours: {
  type: Sequelize.INTEGER,
  validate: { min: 0 }
 }
}) {
 beforeValidate(options) {
 }
}
```

and, bonus, to use `co` (we have build a minimalistic lib `co-wrap-proto`):  

```js
class MyModel extends Sequelize.ModelFactory({
 hours: {
  type: Sequelize.INTEGER,
  validate: { min: 0 }
 }
}) {
 *beforeValidate(options) {
 }
}
MyModel = coWrapProto(MyModel); // Wrap all generator ownProperties into `co.wrap`
```

**Comment by mickhansen** [comment]
Thanks for the suggestions @Verdier, although the ModelFactory stuff seems non-idiomatic

**Comment by mickhansen** [comment]
ES7+ Property Initializers would solve this even better than decorators

**Comment by Verdier** [comment]
That's a point of view @mickhansen, we think decorator are more readable, and allow function decoration. And it's extremely simple to build decorators which work with today Sequelize. Moreover, if you want to use TypeScript, decorators allow you to type the property. But that's another question.

For the ModelFactory, I'm curious to see another readable solution. You could manually decorate the class, but attributes will be after the body:
```js
class MyModel extends Sequelize.Model {}
MyModel.options = { ... };
```

**Comment by mickhansen** [comment]
@Verdier definitely a point of view, and a mute one at this point since we need a solution that also works without Babel :)

Yeah manually adding afterwards would work. I wonder what type of javascript code decorators and property Initializers map to (whether or not it's equivelant to setting a static prop after the fact).

**Comment by Verdier** [comment]
More than my previous arguments we could not use property initialisers since you need to instantiate the class to retrieve the options... Here is where decorators come.

**Comment by mickhansen** [comment]
@Verdier Can you show me a decorator proposal plus what the code maps to in javascript?

**Comment by mickhansen** [comment]
Ah nvm, decorators call a method, so wouldn't necessarily help us.

**Comment by Verdier** [comment]
@mickhansen yep a decorator is a method, what's the problem with that? Here is how we use decorator today (partial introduction, we could have an in depth discussion later):

```js
function attribute() {
        return function (target, key, options) {
            _initializeConfiguration(target);
            target.$config.attributes[key] = options;
        }
}
```
at the end, all the `model` options are in `MyModel.prototype.$config` and accessible without instantiation. Then: `Sequelize.define('myModel', MyModel.prototype.$config)`. That's all!

**Comment by mickhansen** [comment]
The problem is that it does not map nicely to an ES2015 class.
For it to map nicely you'd need to have methods like:

```js
import {
  Model
} from 'sequelize';

class SomeModel extends Model {

}

Model.attribute(SomeModel, {
  type: ...
});
```

etc etc, not sure if that's the nicest syntax :)

**Comment by mickhansen** [comment]
But @Verdier lets move discussion to the more relevant issue here: https://github.com/sequelize/sequelize/issues/4728

**Comment by Verdier** [comment]
@mickhansen I totally agree, that's not a solution for ES6 class. Angular 2 try to solve the problem (they massively use decorators) by providing solution for ES5/6. We could have a look.

**Comment by felixfbecker** [comment]
@mickhansen technically, decorators can be used in ES6. A decorator is nothing but a function that takes a class and puts additional info on it. There is just no syntactic sugar for it there yet. So in ES6 you could either assign the info yourself like @Verdier suggested or just pass it to a function like you do today:
```js
class MyModel extends Sequelize.Model {
  static findByMyAwesomeAttribute() { ... }
  get myCustomAttribute() { ... }
}
seq.define(MyModel, {
  attributes: {
    id: Sequelize.INTEGER,
    name: Sequelize.STRING
  },
  tableName: 'my_model'
})
```
In ES7 or TypeScript you could then use `@tableName()` or `@attributes` decorators

**Comment by mickhansen** [comment]
@felixfbecker If we are going to support ES6 classes it's with decoupling in mind, having to use sequelize.define still would defeat the purpose.

**Comment by felixfbecker** [comment]
@mickhansen define in that case would only be a decorator function that is equivalent to
```js
MyModel.sequelizeOptions = {
  attributes: {
    id: Sequelize.INTEGER,
    name: Sequelize.STRING
  },
  tableName: 'my_model'
}
```
So, in a way, it would be a static property (but not a static method). But if ES at some point gains the ability to define properties, it should be possible to use decorators/annotations on those, like you can with TypeScript today:
```ts
@tableName('my_model')
class MyModel {
  
  @type(Sequelize.INTEGER)
  @autoIncrement
  public id: integer;

  @type(Sequelize.VARCHAR(255))
  public name: string;
}
```

**Comment by mickhansen** [comment]
@felixfbecker I definitely like decorators in the future, but we need decent syntax for ES6 classes first. But please move discussion to the issue dealing with ES6 classes: https://github.com/sequelize/sequelize/issues/4728

**Comment by felixfbecker** [comment]
Sorry for hijacking this issue

**Comment by mickhansen** [comment]
@felixfbecker Not a problem at all! But just in case people are watching the other issue it'd be great to get everyone involved :)

**Comment by overlookmotel** [comment]
To return to original subject of this issue, how about continuing to test against node v0.12 in Travis but (1) making sure those test runs are scheduled after the v4 and v5 test runs, and (2) marking them as allowed to fail in Travis?

That way you'd get the green tick from Travis quicker, but still be able to keep an eye on whether anything is broken on node v0.12.

**Comment by Verdier** [comment]
What about ES6 features?

**Comment by overlookmotel** [comment]
@Verdier Well @mickhansen was saying he'd like v0.12 to remain unofficially supported but to stop testing against it to speed up Travis. I have no idea how support for both v0.12 and ES6 is going to be achieved - I'm just saying that if the aim is to maintain some support for v0.12, we need some indication of when things have got broken.

**Comment by Verdier** [comment]
There is no support for ES6 in node 0.12 and even with `--harmony` it's still very incomplete. One `class` or `=>` somewhere in the code and none of the v0.12 tests will pass...

**Comment by mickhansen** [comment]
We will be doing ES6 through babel, we'll be compiling to code that should still run on Node v0.10.
We just don't want to bother with running CI for it. Having 0.12 as an allowed failure could be a way to go.

**Comment by felixfbecker** [comment]
Should this also mean to drop the bluebird dependency?

**Comment by mickhansen** [comment]
No, ES6 promises are slower and lack features we rely on, and do not necessarily exist in older versions.

**Comment by sushantdhiman** [comment]
Yes please remove other Node and IO versions from CI, +1

**Comment by overlookmotel** [comment]
I think I mentioned in another issue thread I've been wondering if there's a way to speed up the tests runs by running multiple tests in parallel (within the same Travis process). This would speed things up a lot as most of the time is spent waiting for async IO with the database. But I don't know if anyone's already built something like this for mocha.

**Comment by mickhansen** [comment]
@overlookmotel Tricky for our integration tests which all hit the database though, would have to have isolated databases, but i suppose that could actually be doable on Travis.

**Comment by sushantdhiman** [comment]
Can it be closed now as #5918 has been merged to master ? @janmeier @mickhansen

**Comment by overlookmotel** [comment]
Where shall we discuss what ES6 features should be permitted in Sequelize v4 codebase? Or is it basically anything that's supported in node v4?

**Comment by mickhansen** [comment]
https://github.com/sequelize/sequelize/issues/5898

**Comment by mickhansen** [comment]
@overlookmotel i guess basically anything that's supported by v4, but we haven't decided yet.
